### PR TITLE
Cosmos Issue #6240: Port read-my-writes fix from V4

### PR DESCRIPTION
**Pull request by David-Noble-at-work**
See #6240 and https://github.com/Azure/azure-sdk-for-java/pull/6201.

From a v3.X perspective this PR eliminates the read-my-writes test failure and picks up a number of diagnostic improvements that have been introduced to the Direct TCP stack. This PR does not bump the version number. This change will need to be done on a separate PR.

### CODE REVIEW

File: sdk/cosmos/microsoft-azure-cosmos-benchmark/src/main/java/com/azure/data/cosmos/benchmark/ReadMyWriteWorkflow.java [hunk 1]
```diff
@@ -90,12 +95,15 @@ protected void performWorkload(BaseSubscriber<Document> baseSubscriber, long i)
                     throw new IllegalStateException();
             }
         } else {
+
             // will either do
             // a write
-            // a point read for a in memory cached document
+            // a point read for a in memory cached document4
```
Comment (moderakh) [inline sdk/cosmos/microsoft-azure-cosmos-benchmark/src/main/java/com/azure/data/cosmos/benchmark/ReadMyWriteWorkflow.java]: document4 -> document

File: sdk/cosmos/microsoft-azure-cosmos-benchmark/src/main/java/com/azure/data/cosmos/benchmark/ReadMyWriteWorkflow.java [hunk 2]
```diff
@@ -124,8 +132,14 @@ protected void performWorkload(BaseSubscriber<Document> baseSubscriber, long i)
...
 
         concurrencyControlSemaphore.acquire();
+        logger.debug("concurrencyControlSemaphore: {}", concurrencyControlSemaphore);
 
-        obs.subscribeOn(Schedulers.parallel()).subscribe(baseSubscriber);
+        try {
+            obs.subscribeOn(Schedulers.parallel()).subscribe(baseSubscriber);
+        } catch (Throwable error) {
+            concurrencyControlSemaphore.release();
+            logger.error("subscription failed due to ", error);
```
Comment (moderakh) [inline sdk/cosmos/microsoft-azure-cosmos-benchmark/src/main/java/com/azure/data/cosmos/benchmark/ReadMyWriteWorkflow.java]: subscribe should not ever through at least on v2. Have you seen cases of `subscribe` throwing?

if so could you please share the exception stacktrace on the code review comment?

**Comment by kushagraThapar** [comment]
Tests are failing because of emulator breaking change. I remember seeing those failures back in the day when we used daily emulator build in the CI. 
Now despite using public emulator, since these tests are still failing, it could be possible that we need to update those.

File: sdk/cosmos/microsoft-azure-cosmos-benchmark/src/main/java/com/azure/data/cosmos/benchmark/ReadMyWriteWorkflow.java (hunk 1 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/microsoft-azure-cosmos-benchmark/src/main/java/com/azure/data/cosmos/benchmark/ReadMyWriteWorkflow.java]: thanks for finding this. it is now corrected.

File: sdk/cosmos/microsoft-azure-cosmos-benchmark/src/main/java/com/azure/data/cosmos/benchmark/ReadMyWriteWorkflow.java (hunk 2 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/microsoft-azure-cosmos-benchmark/src/main/java/com/azure/data/cosmos/benchmark/ReadMyWriteWorkflow.java]: thanks for finding this. it is now corrected.

**Pull request merged**